### PR TITLE
:unicorn: added color segmentation

### PR DESCRIPTION
**Pull request by tohahn**
Added working color segmentation to robosherlock.

### CODE REVIEW

File: src/segmentation/include/rs/HueClusterComparator.h [hunk 1]
```diff
@@ -0,0 +1,206 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2010-2012, Willow Garage, Inc.
```
Comment (bbferka) [inline src/segmentation/include/rs/HueClusterComparator.h]: Did you implement this yourself or is it really from PCL?

**Comment by tohahn** [comment]
the base code is from pcl. i added the code necessary for hue comparison and removed the code for specifying which points should be compared (labels).

**Comment by bbferka** [comment]
do you think you could make this work on existing hypotheses as well?..as an option in the xml, the input cloud would then be a hypotheses that matches some description.

**Comment by tohahn** [comment]
I think you mean the following: The cloud that should be clustered should
in some cases not be the scene but a robosherlock cluster that we think is
some object. That would be possible, but I need to know what the
description of the hypothis is and how to get the rs cluster based on the
description. Or do you mean something different?

**Comment by bbferka** [comment]
exactly. The description of that hypotheses could be in the query:

`rs::Query q = rs::create<rs::Query>(tcas);`
`cas.getFS("QUERY",` q);`

you can find an example in the RegionFilter. Query has a field called asJson. and you can check there for an ID for example, which needs to match the idx of the cluster....for now this would suffice.

**Comment by tohahn** [comment]
will do.

2017-05-03 14:48 GMT+02:00 Ferenc Balint-Benczedi <notifications@github.com>
:

**Comment by tohahn** [comment]
i have two problems:
1. how do i get the id from the asJson?
2. how do I get the ID from a rs::Cluster?

2017-05-03 16:11 GMT+02:00 Tobias Hahn <morowinger@gmail.com>:

**Comment by bbferka** [comment]
asJson is a json string...you can use a parser to look for an id field. for rs::Cluster you can simply use the index in the vector for now

**Comment by tohahn** [comment]
The color clusterer now clusters for hue and for value. Also, it is possible to discretize values beforehand to avoid spilling. The possibility to cluster only on a hypothesis (id of a preexisting cluster) was also added.

**Comment by bbferka** [comment]
looks ok, but you should not put rapidjson in the project. We don't like introducing dependencies that do not have a debian package. With rapidjson this will be solved once we migrate 16.04. If  this is simply there for the checking of the query for ID, then I suggest replacing it with a simple comparison e.g. asJson == "clusterNR". 

The other option is setting it as an external dependency, looking for rapidjson and conditioning the compilation of the annotator on this:
```cmake
find_package(RapidJSON QUIET)

if(${RapidJSON_FOUND})
  message("Found RapidJson, building XYZ Annotator")
  include_directories(${RAPIDJSON_INCLUDE_DIRS})
endif()
```

**Comment by tohahn** [comment]
the lookup in the json string is now done with standard string functions.

**Comment by bbferka** [comment]
hmm...this is an output that I get:
```
PointCloudColorSegmentation.cpp(159)[processWithLock] Found 1904 clusters for 1228800 points.
````
I should not end up with a point cloud of that size. It seems that you are concatenating point clouds, and keeping old indices as a result, or something similar (see picture).

screenshot from 2017-06-21 12 48 01

**Comment by tohahn** [comment]
the latest commit was formatted with the styling guidelines and fixes the bug with old point cloud information by clearing the clouds before every iteration of the annotator.

**Pull request closed without merging**